src/renderTemplate.tsx

```diff
@@ -30,7 +30,10 @@ export async function renderTemplate(
 
     // Consolidate mutates the `locals` input, so provide a copy or otherwise an
     // empty object if not locals were specified.
-    const html = await compileFn(absoluteFilePath, { ...options.locals })
+    const html = await compileFn(absoluteFilePath, {
+      ...options.locals,
+      cache: true,
```

Comment:
Something similar led to an [incident a while back](https://github.com/artsy/post_mortems/issues/18) (pre stitch) where the cache would persist across users and folks would see someone else logged in (visually). Can you verify these sorts of things don't occur?

Reply:
The docs _seem_ to indicate only things we want to be cached are cached–which are I/O and, from my profiling, parsing of templates. https://github.com/tj/consolidate.js/#caching

Reply:
As long as it can be verified IRL against the linked issue above I'm down to approve; the source-code reference in the incident was my template for Stitch, and I did something similar there as you're doing here, leading to the problem.

Reply:
Yeah we should at the very least verify this in staging. All my browsing through source of consolidate.js and jade, the `cache` option is about caching functions that it generates out of the templates–functions _imply_ that things like locals would be passed in as arguments–and that you should always cache https://github.com/pugjs/pug/issues/320#issuecomment-1994202

Reply:
But it does look very much like that’s what the original change was doing too.

However, the [PR that fixed it at the time](https://github.com/artsy/force/pull/1549) mentions `sd` data being leaked and that usually happens because somebody adds data to the `sd` global singleton. Could that have been it?

Reply:
👍

Reply:
Note https://github.com/artsy/force/pull/1549#discussion_r122014203

Reply:
Hmm, too bad there’s no breadcrumbs (working any longer) about how that was determined.

Reply:
Ok, live blogging as I go through this:

* Ok, on to the first compilation of the jade template into a render function. 
  1. [The entry point `compileFile`](https://github.com/pugjs/pug/blob/1.11.0/jade.js#L308) quickly takes us to [`handleTemplateCache`](https://github.com/pugjs/pug/blob/1.11.0/jade.js#L170).
  1. This will in turn [take us to the compilation of the template function](https://github.com/pugjs/pug/blob/1.11.0/jade.js#L176).
  1. `compile` then [parses/compiles the given template source string into a re-usable function](https://github.com/pugjs/pug/blob/1.11.0/jade.js#L199-L220), which will look something like [this](https://gist.github.com/alloy/2e1ca1886842667f71beadda5f8694b9) (example of `react_redesign.jade`).
  1. Popping back the stack to `handleTemplateCache`, which will then [cache the resulting template function](https://github.com/pugjs/pug/blob/1.11.0/jade.js#L177) using the filename of the template as cache key, before returning it back up the stack to `consolidate.js` for rendering.

* Subsequent invocations of the `consolidate.js` `jade` function will:
  1. Return the cached template function from [Jade’s `handleTemplateCache` function](https://github.com/pugjs/pug/blob/1.11.0/jade.js#L173).
  1. Invoke the template function [with new locals passed in](https://github.com/pugjs/pug/blob/1.11.0/jade.js#L220-L221).

It’s important to note that consolidate calls both [the Jade `compileFile` function](https://github.com/tj/consolidate.js/blob/v.0.14.5/lib/consolidate.js#L308) _and_ [the actual rendering of the template (`tmpl`)](https://github.com/tj/consolidate.js/blob/v.0.14.5/lib/consolidate.js#L309) with the same set of `options`, which is overloaded to both take options _and_ contain local variable definitions.

This “mix objects together” type of design doesn’t allow for the biggest trust into what’s happening–imo it would have been better to separate the two and know for sure that only `{ cache: true }` was actually being sent to the compile function and being included in the scope of the resulting template function.

Having said that, unless the `options` object would contain different keys per invocation, any locals that were erroneously in scope at compile time would be expected to be overridden at render time.

So I am fairly confident in saying that the previous incident must have been due to somebody adding personalised data to the `sharify.data` singleton, rather than the per-request `res.locals` copy.

### Follow-up work

* Seems like we are compiling templates with debugging enabled, going to ensure that that’s not an artefact of my debugging setup and otherwise disable that to improve performance further.

* We are most definitely still using the last **Jade** version when doing SSR, _not_ Pug. I can imagine that using a maintained version could also bring some performance benefits. (Although not necessarily, it could also have regressed.)

* Why do we need to use `consolidate.js` at all? Are we rendering other types of templates? Because getting rid of that dependency would allow us to explicitly compile the template with no local variables and invoke the template function with only locals.

Reply:
Stitch is meant to be used with any templating engine. Consolidate is a common API used across these various renderers; the user still needs to install, say, jade, or ejs, or whatever. I don't think we will gain much refactoring this, is lightweight, and it goes against the purpose of the library. 

As far as upgrading from jade to pug, feel free, but not sure if thats a priority kind of thing. Gone down this road and we've got ten million jade templates. Not sure what kind of benefit this will bring when our new code is React and soon the main shell won't even use Jade for all of the important pages.

Reply:
> Stitch is meant to be used with any templating engine. […] I don't think we will gain much refactoring this, is lightweight, and it goes against the purpose of the library.

Isn’t the purpose of the library to solve a problem _we_ have? I ask, because adding this abstraction layer that follows a lowest common denominator approach means that it makes it one step harder to easily make use/find underlying specific API options (e.g. not compiling debug info).

If we actually have a need for rendering multiple types of templates, then I totally understand the design. If not, however, then I would suggest in the future to refrain from making things more generic than the known users (i.e. only us) need. Staying closer to implementation details often makes it easier to do the right thing, form my experience.

> As far as upgrading from jade to pug […] Not sure what kind of benefit this will bring when our new code is React and soon the main shell won't even use Jade for all of the important pages. But if you feel it's important by all means dive in!

I see, in hindsight this nugget of information would indeed have been a good reason not to go down this path, but from other conversion we had it seemed like you were under the impression that we _were_ in fact using pug and so I made an incorrect assumption that I should have validated with you first. My bad.

Anyways, by now I’ve already done the conversion–which wasn’t too bad, just needs some proper testing.

Reply:
> Isn’t the purpose of the library to solve a problem we have

Not necessarily. Consolidate.js has been around for ever. It's a solid abstraction and common in most implementations for node templating libraries.  It's lightweight, and is simply an API interface. I understand your sentiment about keeping things bare, but I'm not concerned with it getting in the way of anything that we need to do here and I would like to keep it, it's explicitly a part of Stitch's design.

Reply:
I should have been explicit, I meant ‘stitch’ not ‘consolidate’.

Reply:
> Isn’t the purpose of the library to solve a problem we have

Stitch has solved the problem that we have, in multiple ways. So I'm not sure what you mean?